ksqldb-common/src/main/java/io/confluent/ksql/schema/ksql/LogicalSchema.java

```diff
@@ -236,48 +271,97 @@ public String toString(final FormatOptions formatOptions) {
+    }
+
+    if (pseudoColumnVersion >= ROWPARTITION_ROWOFFSET_PSEUDOCOLUMN_VERSION) {
+      builder.add(Column.of(ROWPARTITION_NAME, ROWPARTITION_TYPE, VALUE, valueIndex++));
+      builder.add(Column.of(ROWOFFSET_NAME, ROWOFFSET_TYPE, VALUE, valueIndex++));
+    }
 
+    for (final Column c : key) {
       builder.add(Column.of(c.name(), c.type(), VALUE, valueIndex++));
     }
 
-    if (withPseudoAndKeyColsInValue) {
-      builder.add(Column.of(ROWTIME_NAME, ROWTIME_TYPE, VALUE, valueIndex++));
+    if (windowedKey) {
+      builder.add(
+          Column.of(WINDOWSTART_NAME, WINDOWBOUND_TYPE, VALUE, valueIndex++));
+      builder.add(
+          Column.of(WINDOWEND_NAME, WINDOWBOUND_TYPE, VALUE, valueIndex));
+    }
+
+
+    return new LogicalSchema(builder.build());
+  }
+
+  /**
+   * Rebuilds schema without pseudocolumns or key columns
+   * @return the LogicalSchema created, with the corresponding pseudo and key columns excluded
+   */
+  private LogicalSchema rebuildWithoutPseudoAndKeyColsInValue(final int pseudoColumnVersion) {
+    final ImmutableList.Builder<Column> builder = ImmutableList.builder();
+    addKeyColumnsToKeySchema(builder);
+    addNonPseudoAndKeyColsToValueSchema(builder, pseudoColumnVersion);
+    return new LogicalSchema(builder.build());
+  }
 
-      for (final Column c : key) {
-        builder.add(Column.of(c.name(), c.type(), VALUE, valueIndex++));
+  /**
+   * Adds columns, except for key and pseudocolumns, to a provided builder's value schema
+   * and returns the number of added columns
+   * @param builder the builder to be passed in. Columns that don't qualify as key or pseudocolumns
+   *                will be added
+   * @return the number of columns added. This also functions as a "column index", to pass into
+   * builder.add()
+   */
+  private int addNonPseudoAndKeyColsToValueSchema(
+      final ImmutableList.Builder<Column> builder, final int pseudoColumnVersion) {
+    final Map<Namespace, List<Column>> byNamespace = byNamespace();
+
+    final List<Column> value = byNamespace.get(VALUE);
+
+    int addedColumns = 0;
+    for (final Column c : value) {
+      if (SystemColumns.isSystemColumn(c.name(), pseudoColumnVersion)) {
+        continue;
       }
 
-      if (windowedKey) {
-        builder.add(
-            Column.of(WINDOWSTART_NAME, WINDOWBOUND_TYPE, VALUE, valueIndex++));
-        builder.add(
-            Column.of(WINDOWEND_NAME, WINDOWBOUND_TYPE, VALUE, valueIndex));
+      if (findColumnMatching(withNamespace(Namespace.KEY).and(withName(c.name()))).isPresent()) {
+        continue;
       }
+
+      builder.add(Column.of(c.name(), c.type(), VALUE, addedColumns++));
     }
+    return addedColumns;
+  }
 
-    return new LogicalSchema(builder.build());
+  /**
+   * Adds key columns to key schema of a provided builder
+   * @param builder the builder to be passed in. This builder will have key columns added to it's
```

Comment:
```suggestion
   * @param builder the builder to be passed in. This builder will have key columns added to its
```

Reply:
I've always been SURE it was the contraction and possessive, so I'm legitimately very happy you pointed this out. Maybe it's not a big improvement to the codebase, but it sure is to my English :)